helix-core/src/syntax.rs

```diff
@@ -1589,6 +1721,53 @@ impl<'a> HighlightIter<'a> {
             }
         }
     }
+
+    /// Scans forward to the next capture from whichever layer has the earliest highlight
+    /// boundary, returning highlight-end events or terminating the iterator if there
+    /// are no more captures or highlight-end events.
+    fn scan_to_earliest_capture(&mut self) -> Option<Option<Result<HighlightEvent, Error>>> {
```

Comment:
Do we really need so many layers here? `Option` and `Result`

Reply:
It does look really badly nested 😄

It's really `Option<Option<Self::Item>>` using the Iterator implementation's item. It does need the two nested `Option`s though:

* `None` for the case where we've already scanned to the next capture and can continue
* `Some(Some(event))` for emitting HighlightEnds for captures that have ended when handling the next capture
* `Some(None)` for terminating the iterator when it's at the end of all captures